EOS-15892: Plugable framework

ajay-paratmandali opened the pull request:
Signed-off-by: Ajay Paratmandali <ajay.paratmandali@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    Your Problem statement here...
EOS-15892
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes, Tested with RPM, This patch depend on and tested with https://github.com/Seagate/cortx-ha/pull/300
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Your Problem decription here...
Cluster should be managed by central framework.
  </code>
</pre>
## Solution
<pre>
  <code>
    Your Problem solution here...
Provide cluster manager and pluggable framework to manage cluster as per LLD.
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Unit Testing details here...
Yes, Test case provided in PR
  </code>
</pre>

ajaysrivas commented:
@ajay-paratmandali, could you please update the information in PR template.

ajay-paratmandali commented:
@ajaysrivas Updated PR template.
This PR is depend on https://github.com/Seagate/cortx-ha/pull/300

ajaysrivas commented on ha/const.py:
```diff
@@ -89,3 +90,8 @@
+# Cluster manager
+CM_CONTROLLER_INDEX="controller_interface"
+CM_CONTROLLER_SCHEMA="{}/controller_interface.json".format(CONFIG_DIR)
+CM_ELEMENT=["cluster", "node", "service"]
```
storageset should also be there?
It's a logical entity though.

ajay-paratmandali replied:
Done

ajaysrivas commented on ha/core/cluster/cluster_manager.py:
```diff
@@ -16,16 +16,19 @@
+from ha.core.controller.element_controller_factory import ElementControllerFactory
+from ha.core.error import ClusterManagerError
 
 class ClusterManager:
```
We need to review names of functions.
Also, the functionality regarding storageset. Probably it's missing in HLD too. Let's discuss this after I am done with other PR too.

ajay-paratmandali replied:
Added storageset element and its operation

ajay-paratmandali replied:
Done

indrajitzagade commented on ha/core/cluster/cluster_manager.py:
```diff
@@ -294,73 +275,61 @@ def status(self):
         Manage cluster operation
         """
-        self._execute = SimpleCommand()
+        self._cluster_type = Conf.get(const.HA_GLOBAL_INDEX, "CLUSTER_MANAGER.cluster_type")
```
Do we need any new fields in the confstore?

ajay-paratmandali replied:
Yes we need to detect env is it vm or hw
depending on this need to perform stonith operation.

indrajitzagade replied:
Please list the new fields here so that we can add them to the sheet from provisioner.

ajay-paratmandali replied:
in config
  "corosync-pacemaker": {
    "cluster_name": "cortx_cluster",
    "user": "hacluster",
    "secret": ""
}
Here corosync-pacemakerconsidered as type

indrajitzagade replied:
ok, these are already present.

azheregelya commented on ha/setup/ha_setup.py:
```diff
@@ -230,12 +247,33 @@ def __init__(self, args):
+        if "VM" == node_type.upper():
+            Conf.set(self._ha_conf_index, "CLUSTER_MANAGER.env", node_type.upper())
+        else:
+            # TODO: check if any env available other than vm, hw
```
TODO missed or left intentionally?

ajay-paratmandali replied:
it is intentionally. Currently i checked all possible value and only provided is VM any there can be multiple env available for future so added TODO

ajay-paratmandali commented on ha/core/cluster/cluster_manager.py:
```diff
@@ -294,73 +275,61 @@ def status(self):
+            output = {"rc": 1, "status":"", "err": f"Cluster manager failed. Error: {e}"}
+            return output
+
+    def _responce(self, rc: int, status: str, err: str):
```
It is removed as no longer needed

ajay-paratmandali commented on ha/core/cluster/cluster_manager.py:
```diff
@@ -294,73 +275,61 @@ def status(self):
-            action (string): action taken on cluster
             args (dictonery): parameteter
-            output (object): Store output
+            responce (object): Store output
```
This code is updated no longer needed
as current api are cm.cluster_controller.start()

ajay-paratmandali commented on ha/core/cluster/cluster_manager.py:
```diff
@@ -294,73 +275,61 @@ def status(self):
-        raise HAUnimplemented("Cluster shutdown is not supported.")
+        Args:
+            rc (int): Return code.
+            output (str): Output
```
Removed as not needed

ajay-paratmandali commented:
@shindeam @ajaysrivas @indrajitzagade @azheregelya Commenting for review

ajay-paratmandali commented:
@ajaysrivas 
To avoid dublicate class and conflict
removed ElementController from this pr and
added note in https://github.com/Seagate/cortx-ha/pull/300/files ElementController
